src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1405,6 +1450,258 @@ private void WritePackageDependencies()
                 }
             }
 
+            // This implementation was taken from PreprocessPackageDependenciesDesignTime.ExecuteCore
+            // with some minor modifications to use WriteItem instead of saving it into an array.
+            private void WritePackageDependenciesDesignTime()
+            {
+                List<ITaskItem> _packageDependencies = new List<ITaskItem>();
+                GetPackageDependencies();
+
+                List<ITaskItem> _packageDefinitions = new List<ITaskItem>();
+                GetPackageDefinitions();
+
+                var implicitPackageReferences = GetImplicitPackageReferences(_task.DefaultImplicitPackages);
+
+                // We have two types of data:
+                //
+                // 1) "PackageDependencies" which place a package in a given target/hierarchy
+                // 2) "PackageDefinitions" which provide general metadata about a package
+                //
+                // First, we scan PackageDependencies to build the set of packages in our target.
+
+                var allowItemSpecs = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+                foreach (var dependency in _packageDependencies)
+                {
+                    if (dependency.HasMetadataValue(MetadataKeys.ParentPackage))
+                    {
+                        // ignore non-top-level packages (those with ParentPackage)
+                        continue;
+                    }
+
+                    var target = dependency.GetMetadata(MetadataKeys.ParentTarget);
+
+                    if (!StringComparer.OrdinalIgnoreCase.Equals(target, _task.TargetFramework))
+                    {
+                        // skip dependencies for other targets
+                        continue;
+                    }
+
+                    allowItemSpecs.Add(dependency.ItemSpec);
+                }
```

Comment:
This loop should become unnecessary with other changes suggested. We should create a list of dependencies that doesn't have any of the ones that would be filtered out here.

Reply:
done

Reply:
removed in the code cleanup